### PR TITLE
Feature INBA-535 - AssignTask Current Step

### PR DESCRIPTION
**Pull request by jmfadeley**
Round 2: Way better implementation.

To Test:
**Explanation: Earlier-in-the-process tasks assigned after later tasks should supersede the latter.**
Create a new project, with users, usergroups, a few subjects and a few stages, preferably 3.
Assign a task to stage 3.
Check the `ProductUOA` table and ensure that a new entry is added, which includes a `currentStepId`.
Assign another task on the same subject as the previous one.
Check the `ProductUOA` table and ensure that a prior entry has changed to an earlier (less) `currentStepId`.

**Explanation: The earlier task should supersede later tasks assigned after.**
On another subject, add a task to stage 1. 
Check the `ProductUOA` table for a new entry.
On the same subject as the prior entry, assign another task to either stage 2 or 3.
Check the `ProductUOA` table and ensure that the entry has not changed.

**Explanation: Later tasks should not overwrite or restart the tasks from stage 1.**
On the final subject, add a task to stage 1 and another to stage 2. 
Check the `ProductUOA` table for a new entry. Manually update it so `currentStepId` is for stage 2 (probably the current id +1).
Assign a task to stage 3
Check the `ProductUOA` table and ensure that the entry has **not** changed.

Note that the project now sets the product to be active on creation. This is necessary as an inactive product does not allow workflow entry/advancement.

### CODE REVIEW

File: backend/app/controllers/tasks.js [hunk 1]
```diff
@@ -350,3 +375,73 @@ function* checkTaskData(req) {
         yield * common.checkDuplicateTask(req, req.body.stepId, req.body.uoaId, req.body.productId);
     }
 }
+
+function* updateCurrentStepId(req, insertedTaskId) {
+    var thunkQuery = req.thunkQuery;
+
+    // var essenceId = yield * getEssenceId(req, 'Tasks');
```
Comment (Perry5) [inline backend/app/controllers/tasks.js]: Should we delete this if it's not needed?

File: backend/app/controllers/tasks.js [hunk 2]
```diff
@@ -350,3 +375,73 @@ function* checkTaskData(req) {
...
+    var product = yield * common.getEntity(req, req.body.productId, Product, 'id');
+
+    //TODO: Get survey from survery service if needed
+    // var survey = yield * common.getEntity(req, product.surveyId, Survey, 'id');
```
Comment (Perry5) [inline backend/app/controllers/tasks.js]: Should we delete these comments? If the need arises later, we will probably know where it needs to go.

File: backend/app/controllers/tasks.js [hunk 3]
```diff
@@ -350,3 +375,73 @@ function* checkTaskData(req) {
...
+            .join(Task).on(Task.stepId.equals(WorkflowStep.id))
+        )
+        .where(Task.productId.equals(req.body.productId))
+        .and(Task.uoaId.equals(req.body.uoaId))
+        .order(WorkflowStep.position);
+    result = yield thunkQuery(minStepPositionQuery);
+
+    if (!_.first(result)) {
+        debug('Not found min step position for productId `' + req.body.productId + '`');
+        return null;
```
Comment (Perry5) [inline backend/app/controllers/tasks.js]: Do we want to return null here or just throw an error? 
`throw new HttpError(403, 'Not found min step position for productId `' + req.body.productId + '`')`

File: backend/app/controllers/tasks.js (hunk 2 above)
Comment (jmfadeley) [reply backend/app/controllers/tasks.js]: Might as well.

File: backend/app/controllers/tasks.js (hunk 3 above)
Comment (jmfadeley) [reply backend/app/controllers/tasks.js]: Sure.

Comment (jmfadeley) [reply backend/app/controllers/tasks.js]: Done.

**Pull request merged**